docs/debugging-s2i.md

```diff
@@ -0,0 +1,111 @@
+
+       $ ~/sandbox/ruby sti build --loglevel=5 -e "MAVEN_ARGS='-P some-repo,some-other-repo'" file:///home/mfojtik/sandbox/ruby openshift/ruby-20-centos7 foo
+       	 I0519 14:20:51.410089 17228 util.go:34] Using mi@mifo.sk credentials for pulling openshift/ruby-20-centos7
+     	 I0519 14:20:51.410222 17228 main.go:239] An error occurred: malformed env string: some-other-repo'
+
+To deal with this behavior, s2i by default will look for environment variables in the file `.sit/environment` in your source repository.  You can also point to a separate file with environment variable settings with the -E option.
+With both approaches, leveraging a file instead of the command line allows for the values of environment variable to contain commas.  Environment variable processing is described in the [README](https://github.com/openshift/source-to-image#anatomy-of-a-builder-image) as well.
+
+With the above example, whichever file you leverage for environment variables would have this line:
+
+
+```
+MAVEN_ARGS='-P some-repo,some-other-repo'
+```
+
+
+The S2I Scripts You Have Supplied
+--------------
+
+First, a basic reminder:  you should verify your scripts have excutable permissions.
+
+Then, a few of the trickier obstacles that have arisen for users in the past:
+
+####Interfacing With `tar`
+
+As noted [here](https://github.com/openshift/source-to-image/blob/master/docs/builder_image.md#save-artifacts), if you employ a `save-artifacts` script, that script is responsible for properly handling the sending and/or receiving of tar streams.  Issues with `tar` stream processing could result in messages like:
+
+	      W0720 21:07:52.145257 04204 sti.go:131 Clean build will be performed because of error saving previous build artifacts
+	      E0720 21:07:52.145263 04204 sti.go:133] ERROR: timeout waiting for tar stream
+
+
+Review the example [here](https://github.com/gabemontero/source-to-image/blob/master/docs/builder_image.md#save-artifacts), and perhaps revisit the tar man pages or the bash user manual, to help address any problems.
+
+
+####Dowloading / Finding The Scripts
+
+Per this discussion point  [here](https://github.com/openshift/source-to-image/blob/master/docs/builder_image.md#sti-scripts), if your s2i related scripts are already placed in the image, but their locations are not indicated by one of the provided means, attempts to download the scripts will fail.
+
+   	E0720 21:08:37.166063 04204 main.go:328] An error ocurred: scripts inside the image:  image:///path/to/image
+
+Properly reference the script location with the methods described in the above link.
+
+
+####ONBUILD
+
+At various points when executing ONBUILD instructions, if those instructions result in the need for root user access, and the s2i builder has not been invoked to allow
+it to run as root, you will get the expected permission errors for such scenarios.
```

Comment:
no, you can run an image as root despite not being root yourself.  container isolation means (in theory) this is safe.
"docker run -it -u 0 myimage /bin/bash"

and by default an image runs as whatever user the last "USER" line in the Dockerfile indicates.

And the RUN/etc commands are going to run as whatever USER line preceded them in the Dockerfile.

ONBUILD commands will, I think, run as whatever the last USER statement indicates also.

So if you define a dockerfile like:
FROM foo
USER root
RUN some-root-operation
ONBUILD some-root-operation
USER joe

you will be able to docker build the image, but when you then reference that image in another Dockerfile, things will fail because the ONBUILD commands will execute as "joe" and not "root".